README.md
# OpenCloset-Cron-Visitor #

[![Build Status](https://travis-ci.org/opencloset/OpenCloset-Cron-Visitor.svg?branch=v0.3.12)](https://travis-ci.org/opencloset/OpenCloset-Cron-Visitor)

예약/방문/미방문/대여 수와 이벤트 수치를 통계용 table 인 visitor 에 넣어주는 cronjob

- 일일 온/오프라인 방문자수를 계산 (AM 00:05)
- 일일 이벤트 방문/미방문수를 계산 (AM 00:10)

## Build docker image ##

    $ docker build -t opencloset/cron/visitor .


Changes
# Change Log
The format is based on [Keep a Changelog](http://keepachangelog.com/)

## [Unreleased]

## [v0.3.12] Thu, 03 Oct 2019 17:22:58 +0900
### Fixed
- 같은 이름의 n회차 이벤트 통계가 중복으로 쌓이는 것을 수정 (#21)

## [v0.3.11] - Tue, 15 Jan 2019 14:27:11 +0900

### Changed
- 서울시 이벤트 역시 일반적인 이벤트처럼 다룸 (#20)

## [v0.3.10] - Fri, 11 Jan 2019 00:09:40 +0900

### Fixed
- 잘못된 통계수집 쿼리 변경 (#19)

## [v0.3.9] - Sun, 16 Dec 2018 00:35:37 +0900

### Changed
- rename seoul-2018 to seoul-2018-1

## [v0.3.8] - Sat, 15 Dec 2018 20:21:14 +0900

### Removed
- 이벤트가 끝난 linkstart 통계수집을 삭제 (#17)

### Changed
- 반복되던 통계수집 worker 를 `event` 테이블을 활용해서 대체 함 (#17)

## [v0.3.7] - Wed, 29 Aug 2018 04:42:59 +0900

### Added
- yongbin201801 통계 수집 (#15)

## [v0.3.6] - Wed, 20 Jun 2018 14:24:20 +0900

### Added
- goyang201801 통계 수집 (#13)

## [v0.3.5] - Fri May 4 17:08:30 2018 +0900

### Added
- travis-ci 에서 docker image 를 build

## [v0.3.4] - Tue, 01 May 2018 23:28:57 +0900

### Added
- samsunglife201801 통계 수집 (#11)
- gwangju201801 통계 수집 (#9)

## [v0.3.3] - Fri, 23 Feb 2018 14:35:34 +0900

### Changed
- 안양시 청년옷장 이벤트명이 `anyangyouth` 에서 `anyang201801` 로 변경됨

## [v0.3.2] - Thu, 22 Feb 2018 15:10:31 +0900

### Added
- 군포시 이벤트의 통계를 수집 (#7)

## [v0.3.1] - Mon, 19 Feb 2018 16:03:19 +0900

### Changed
- 취업날개 쿠폰명이 `seoul-2017` 에서 `seoul-2018` 로 변경됨

## [v0.3.0] - Thu, 02 Nov 2017 15:37:07 +0900

### Added
- Add Dockerfile

## [v0.2.4] - Wed, 18 Oct 2017 16:48:46 +0900

### Added
- 한신대학교 이벤트의 통계를 수집

## [v0.2.3] - Tue, 01 Aug 2017 14:32:13 +0900

### Added
- 안양시 청년옷장 이벤트의 통계를 수집

## [v0.2.2] - Wed, 26 Jul 2017 19:10:11 +0900

### Changed
- 취업날개 연장하기 버튼 데이터를 일일 통계에 더함 (#6)

## [v0.2.1] - Tue, 13 Jun 2017 13:07:02 +0900
### Fixed
- linkstart 통계 수집 오타 수정

## [v0.2.0] - Mon, 12 Jun 2017 19:05:24 +0900
### Changed
- 온라인 통계도 수집

## [v0.1.4] - Thu, 11 May 2017 18:08:47 +0900
### Added
- 일일 인천광역시 일자리정책과 이벤트 방문/미방문수를 계산
- 관악고용센터, 십시일밥, 해피빈캠페인의 통계를 수집
- 관악고용센터 통계 수집용 `event_gwanak` 함수 추가
- 십시일밥 통계 수집용 `event_10bob` 함수 추가
- 해피빈캠페인 통계 수집용 `event_happybean` 함수 추가

v0.1.3    Sun, 30 Apr 2017 14:06:10 +0900
  - import `event_linkstart` at bin/opencloset-cron-visitor.pl

v0.1.2    Sat, 29 Apr 2017 02:52:41 +0900
  - linkstart 가 아닌 취업날개 통계를 수집하는 오타 수정

v0.1.1    Fri, 28 Apr 2017 19:09:20 +0900
  - linkstart 통계 수집 worker 를 등록

v0.1.0    Fri, 28 Apr 2017 17:39:26 +0900
  - 취업날개 통계 수집을 좀 더 정확하게 함
  - linkstart 통계를 수집 (#4)

v0.0.4    Wed, 12 Apr 2017 11:50:49 +0900
  - 취업날개 이벤트 일일 방문자 수 계산방식을 변경

v0.0.3    Tue, 11 Apr 2017 17:36:41 +0900
  - 온라인 일일 대여자 수를 구하는 함수와 worker 를 등록

v0.0.2    Tue, 04 Apr 2017 00:37:24 +0900
  - 취업날개 방문자 수를 구하는 함수와 worker 를 등록

v0.0.1    Mon, 03 Apr 2017 14:57:30 +0900
  - initial release


lib/OpenCloset/Cron/Visitor.pm
package OpenCloset::Cron::Visitor;

require Exporter;
@ISA       = qw( Exporter );
@EXPORT_OK = qw(
    event_common
    visitor_count
);

use OpenCloset::Constants::Status qw(
    $NOT_VISITED
    $RESERVATED
    $RENTAL
    $RENTABLE
    $RETURNED
    $CHOOSE_CLOTHES
    $CHOOSE_ADDRESS
    $PAYMENT
    $PAYMENT_DONE
    $WAITING_DEPOSIT
    $PAYBACK
);

use utf8;
use strict;
use warnings;

=encoding utf8

=head1 NAME

OpenCloset::Cron::Visitor - 방문자수와 관려된 cronjob

=head1 SYNOPSIS

    perl bin/opencloset-cron-visitor.pl /path/to/app.conf

=head1 DESCRIPTION

=over

=item *

일일 방문자수를 계산 (AM 00:05)

=item *

일일 이벤트 방문/미방문수를 계산 (AM 00:10)

=back

=head1 METHODS

=head2 visitor_count( $schema, $date )

일별 방문자 수

=cut

sub visitor_count {
    my ( $schema, $date ) = @_;
    return unless $date;

    my $rs = $schema->resultset('Order')->search( { online => 0 }, { join => 'booking' } )
        ->search_literal( 'DATE(`booking`.`date`) = ?', $date->ymd );

    my %visitor = (
        male   => { reserved => 0, visited => 0, unvisited => 0, rented => 0, bestfit => 0, },
        female => { reserved => 0, visited => 0, unvisited => 0, rented => 0, bestfit => 0, },
    );

    while ( my $order = $rs->next ) {
        my $user      = $order->user;
        my $user_info = $user->user_info;
        next unless $user_info;

        my $gender = $user_info->gender;
        next unless $gender;

        ++$visitor{$gender}{reserved};
        ++$visitor{$gender}{rented} if $order->rental_date;
        if ( $order->status_id =~ m/^1[24]$/ ) {
            ++$visitor{$gender}{unvisited};
        }
        else {
            ++$visitor{$gender}{visited};
        }

        if ( $order->bestfit ) {
            ++$visitor{$gender}{bestfit};
        }
    }

    $rs = $schema->resultset('Order')->search(
        {
            online      => 1,
            rental_date => {
                -between => [
                    $date->clone->truncate( to => 'day' )->datetime,
                    $date->clone->add( days => 1 )->subtract( seconds => 1 )->datetime
                ]
            }
        },
        undef
    );

    my %online = (
        male   => { rented => 0 },
        female => { rented => 0 },
    );

    while ( my $order = $rs->next ) {
        my $user      = $order->user;
        my $user_info = $user->user_info;
        next unless $user_info;

        my $gender = $user_info->gender;
        next unless $gender;

        ++$online{$gender}{rented};
    }

    return {
        offline => \%visitor,
        online  => \%online,
    };
}

=head2 visitor_count_online( $schema, $date )

온라인 일별 대여자 수

=cut

sub visitor_count_online {
    my ( $schema, $date ) = @_;
    return unless $date;

    my $from = $date->clone->truncate( to => 'day' );
    my $to = $from->clone;
    $to->set( hour => 23, minute => 59, second => 59 );
    my $rs = $schema->resultset('Order')->search(
        {
            online      => 1,
            rental_date => { -between => [ $from->datetime(), $to->datetime() ] }
        },
        undef
    );

    my %visitor = (
        male   => { rented => 0 },
        female => { rented => 0 },
    );

    while ( my $order = $rs->next ) {
        my $user      = $order->user;
        my $user_info = $user->user_info;
        next unless $user_info;

        my $gender = $user_info->gender;
        next unless $gender;

        ++$visitor{$gender}{rented};
    }

    return \%visitor;
}

=head2 event_common( $schema, $date, $event )

C<$name> 이벤트 방문자 수

=cut

sub event_common {
    my ( $schema, $date, $event ) = @_;
    return unless $date;
    return _event_daily( $schema, $date, $event );
}

=head2 _event_daily($schema, $date, $event)

=cut

sub _event_daily {
    my ( $schema, $date, $event ) = @_;
    return unless $date;

    my %visitor = (
        male   => { visited => 0, unvisited => 0 },
        female => { visited => 0, unvisited => 0 },
        10     => { visited => 0, unvisited => 0 },
        20     => { visited => 0, unvisited => 0 },
        30     => { visited => 0, unvisited => 0 },
    );

    my $year = $date->year;
    my $rs   = $schema->resultset('Order')->search(
        {
            'me.online'     => 0,
            'me.status_id'  => { 'not in' => [ $NOT_VISITED, $RESERVATED ] },
            'coupon.status' => 'used',
            'event.id'      => $event->id,
        },
        {
            select => [
                'user_info.gender',
                'user_info.birth',
            ],
            as => [
                'gender',
                'birth'
            ],
            join => [
                'booking',
                { coupon => 'event' },
                { user   => 'user_info' },
            ]
        }
    )->search_literal( 'DATE(`booking`.`date`) = ?', $date->ymd );

    while ( my $row = $rs->next ) {
        my $gender = $row->get_column('gender');
        my $birth  = $row->get_column('birth');
        next unless $gender;
        next unless $birth;

        $visitor{$gender}{visited}++;

        my $age = int( ( $year - $birth ) / 10 ) * 10;
        $visitor{$age}{visited}++;
    }

    $rs = $schema->resultset('Order')->search(
        {
            'me.online'     => 0,
            'me.status_id'  => { -in => [ $NOT_VISITED, $RESERVATED ] },
            'coupon.status' => 'reserved',
            'event.id'      => $event->id,
        },
        {
            select => [
                'user_info.gender',
                'user_info.birth',
            ],
            as => [
                'gender',
                'birth'
            ],
            join => [
                'booking',
                { coupon => 'event' },
                { user   => 'user_info' }
            ]
        }
    )->search_literal( 'DATE(`booking`.`date`) = ?', $date->ymd );

    while ( my $row = $rs->next ) {
        my $gender = $row->get_column('gender');
        my $birth  = $row->get_column('birth');
        next unless $gender;
        next unless $birth;

        $visitor{$gender}{unvisited}++;

        my $age = int( ( $year - $birth ) / 10 ) * 10;
        $visitor{$age}{unvisited}++;
    }

    ## online
    $rs = $schema->resultset('Order')->search(
        {
            'me.online'      => 1,
            'me.rental_date' => {
                -between => [
                    $date->datetime,
                    $date->clone->add( days => 1 )->subtract( seconds => 1 )->datetime
                ]
            },
            'me.status_id' => { -in => [ $RENTAL, $RETURNED ] }, # 대여중 혹은 반납
            'coupon.status' => 'used',
            'event.id'      => $event->id,
        },
        {
            select => [
                'user_info.gender',
                'user_info.birth',
            ],
            as => [
                'gender',
                'birth'
            ],
            join => [
                { coupon => 'event' },
                { user   => 'user_info' }
            ]
        }
    );

    my %online = (
        male   => { visited => 0, unvisited => 0 },
        female => { visited => 0, unvisited => 0 },
        10     => { visited => 0, unvisited => 0 },
        20     => { visited => 0, unvisited => 0 },
        30     => { visited => 0, unvisited => 0 },
    );

    while ( my $row = $rs->next ) {
        my $gender = $row->get_column('gender');
        my $birth  = $row->get_column('birth');
        next unless $gender;
        next unless $birth;

        $online{$gender}{visited}++;

        my $age = int( ( $year - $birth ) / 10 ) * 10;
        $online{$age}{visited}++;
    }

    $rs = $schema->resultset('Order')->search(
        {
            'me.online'      => 1,
            'me.rental_date' => undef,
            'me.wearon_date' => $date->datetime(),
            'coupon.status'  => 'used',
            'event.id'       => $event->id,
        },
        {
            select => [
                'user_info.gender',
                'user_info.birth',
            ],
            as => [
                'gender',
                'birth'
            ],
            join => [
                { coupon => 'event' },
                { user   => 'user_info' }
            ]
        }
    );

    while ( my $row = $rs->next ) {
        my $gender = $row->get_column('gender');
        my $birth  = $row->get_column('birth');
        next unless $gender;
        next unless $birth;

        $online{$gender}{unvisited}++;

        my $age = int( ( $year - $birth ) / 10 ) * 10;
        $online{$age}{unvisited}++;
    }

    return {
        offline => \%visitor,
        online  => \%online,
    };
}

1;

=head1 COPYRIGHT and LICENSE

The MIT License (MIT)

Copyright (c) 2017 열린옷장

=cut
